### DISCUSSION TITLE
Example of Virtualenv Discover Selection?

### DISCUSSION THREAD

**Discussion by lanshark** [tox-dev/tox#2953] [Q&A] [answered]
## What's the problem this feature will solve?

I would like to use either asdf or pyenv to run tests with multiple versions of python.

## Describe the solution you'd like

Provide a decent example of the implementation of the python discovery process, which is extremely poorly documented.  

Both the tox-asdf and tox-pyenv plugins no longer support version 4.  The only extant implementation appears to be tox-pdm, but its not clear this implementation works in a similar manner.

## Alternative Solutions

None - I have already forked both tox-asdf and tox-pyenv, and am willing to do the work to bring them up to support 4.0 - in fact, I have some parts of that work completed. But the discovery mechanism is not well documented. 

## Additional context

Any suggestions, or links to any package which supports the python virtualenv discovery mechanism would be helpful.

**Comment by masenf** [comment, marked as answer]
Have you considered [virtualenv-pyenv](https://github.com/un-def/virtualenv-pyenv)?

It seems to replace the functionality of `tox-pyenv`, but works for virtualenvs outside of tox as well; you could probably depend on it in your tox environment by adding the plugin to the [`requires` list](https://tox.wiki/en/latest/config.html#core)

**Reply by gaborbernat** [reply]
This. tox no longer handles this and instead delegates to virtualenv, so a plugin for that is recommended.

**Reply by lanshark** [reply]
I am gonna look at source for it, and possibly make an asdf version.

**Reply by lanshark** [reply]
Have you considered putting a link to this in the documentation?

**Reply by gaborbernat** [reply]
https://virtualenv.pypa.io/en/latest/extend.html#python-discovery

**Reply by lanshark** [reply]
I meant to un-def/virtualenv-pyenv

**Comment by lanshark** [comment]
Announcing the availability of virtualenv-asdf and tox-asdf-redux to support using asdf for multiple python version testing with Tox 4.

https://pypi.org/project/tox-asdf-redux/
https://pypi.org/project/virtualenv-asdf/

Thanks to un-def for showing the way!
